### ISSUE TITLE
Add launcher option to disable auto build during maven builds

### ISSUE THREAD

**Issue by dstango** [eclipse-m2e/m2e-core#1318]
I like to have auto build on, as it's very to have things readily compiled most of the time during development. Yet during a maven build (especially combined with auto-refresh of the workspace) the auto-build and the maven build easily interfere with each-other. 
Most of my colleagues thus turn off auto-build, which results in habits like only doing time consuming full maven builds, accepting having many error messages hanging around (that an incremental build would have resolved automatically), thus ignoring the eclipse problems list, etc. All-in-all not a good user experience.

Some years ago I stumbled upon a little plugin that I found very handy:
    https://github.com/gayanper/eclipse-build-option-plugin
    https://gayanper.github.io/eclipse-build-option-plugin/

It allows to automatically turn of the auto-build during maven execution, thus no interference and practically the best of both worlds.

Unfortunately the update site is dead, but I was wondering if it wouldn't be a good idea to provide such a feature by default, or probably ask @gayanper if he would be willing to contribute it.

Or maybe there's another way to handle the described situation that I'm not aware of. In that case: please enlighten me :-).

**Comment by laeubi** [comment]
@dstango yes that would be very good, I also stumbled over this problem, if we can solve this withing m2e that's perfectly fine.

Instead of disabling the autobuild, you can take a look at 

https://github.com/eclipse-m2e/m2e-core/blob/aff5a4eae39013a77b7c473ace66e22a2b5642b3/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/builder/MavenBuilderImpl.java#L185

so one could skip any delta of a project that is currently build.

**Comment by dstango** [comment]
So you're saying this could be the place to check: if a resource is located in a project, that has a currently running maven build, then just don't set hasRelevantDelta to true?

**Comment by laeubi** [comment]
Yes the method can return `false`, maybe one can even perform the check earlier, but somewhere around that method is a good place to check, to make it more semantic, one might even want to have a new method "hasRunningMavenBuild" or something like that.

**Comment by dstango** [comment]
as soon as I have a running me2-build that might actually a good place for me to play around, as the scope of change seems probably small enough for a start. ;-)

**Comment by dstango** [comment]
@laeubi So I thought this might be a good place to start ;-)

I'm thinking along the lines of adding something to MavenBuilderImple like this:
```
    AtomicBoolean hasRelevantDelta = new AtomicBoolean();
    resourceDelta.accept(delta -> {
      IResource resource = delta.getResource();
      if(resource instanceof IFile) {
        IPath fullPath = delta.getFullPath();
        if (isCurrentlyBuiltByMavenLauncher(fullPath)) {
          return false;
        }
```

Yet I think I need some hint how to find out which launcher is currently active, and then see which is the corresponding base directory. Any hint on where to find that?

**Comment by dstango** [comment]
Here's the result of some guess-perimentation:
```
  private boolean isCurrentlyBuiltByMavenLauncher(IPath recourcePath) {
    for(ILaunch launch : DebugPlugin.getDefault().getLaunchManager().getLaunches()) {
      if(!launch.isTerminated()) {
        ILaunchConfiguration launchConfig = launch.getLaunchConfiguration();
        if(MavenLaunchConstants.LAUNCH_CONFIGURATION_TYPE_ID.equals(launchConfig.getType().getIdentifier())) {
          Path launchPath = new Path(launchConfig.getAttribute(MavenLaunchConstants.ATTR_POM_DIR, "/dev/nul"));
          if(launchPath.isPrefixOf(recourcePath)) {
            return true;
          }
        }
      }
    }
    return false;
  }
```

@laeubi can you let me know if I'm on the right track, here?

**Comment by laeubi** [comment]
This looks like a good start :+1: 

There might be some pitfalls for special configurations e.g. if the pom that is run includes a module in a different level, but we can first start with the simplest case. Otherwise one needs to determine the full set of modules when executing a pom build an store that in the launch, an alternative would be to query the `IMavenProjectFacade` and map the launch to a n m2e project.

**Comment by laeubi** [comment]
@dstango do you plan to further work on this?

**Comment by dstango** [comment]
@laeubi thanks for following up. I intended to continue working on this, yet I got stuck with some basic issues in how to actually manually explore my changes. Please see here: https://github.com/eclipse-m2e/m2e-core/issues/1190#issuecomment-1476413862
Any hint on this?

**Comment by laeubi** [comment]
I have now reworked Tycho so it hopefully is more lenient about line endings in general:
- https://github.com/eclipse-tycho/tycho/pull/2302

about the IDE setup @HannesWell can maybe help best here.

**Comment by to-fr** [comment]
Hi @dstango , are you still working on this? :-)

**Comment by dstango** [comment]
> Hi @dstango , are you still working on this? :-)

Well, as stated here and in the mentioned issue I got stuck in not knowing how to actually try out my changes on a basic level. As I already tried to get me head around it without success, my work on this is stalled, unless someone can give me a hand.

**Comment by laeubi** [comment]
@dstango can you try with the latest source tree and tell what is not working? I think @HannesWell is working with windows as well and seems to get it working.

**Comment by dstango** [comment]
@laeubi I'll see what I can do, but I'm very time constrained right now, as my little newborn is quite an attention and time magnet ... so I can't promise much right now.

Anyhow: actually I was not so much stuck with the build any more (using your tips and workarounds), but was kind of lost how to start eclipse so I could try out my function. I wrote what I tried in the referenced comment, probably a bit further down.
